File: Cpp/Arrays & Hashing/valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        return isValidRow(board) && isValidColumn(board) && isValidSubgrid(board);
    }

    bool isValidRow(vector<vector<char>>& board) {
        for (vector<char> row : board) {
            if (!isValidUnit(row)) {
                return false;
            }
        }
        return true;
    }
    
    bool isValidColumn(vector<vector<char>>& board) {
        for (int i = 0; i < 9; i++) {
            vector<char> column;
            for (int j = 0; j < 9; j++) {
                column.push_back(board[j][i]);
            }

            if (!isValidUnit(column)) {
                return false;
            }
        }
        return true;
    }

    bool isValidSubgrid(vector<vector<char>>& board) {
        for (int i = 0; i < 9; i += 3) {
            for (int j = 0; j < 9; j += 3) {
                vector<char> subgrid;
                for (int x = i; x < i + 3; x++) {
                    for (int y = j; y < j + 3; y++) {
                        subgrid.push_back(board[x][y]);
                    }
                }
                if (!isValidUnit(subgrid)) {
                    return false;
                }
            }
        }
        return true;
    }

    bool isValidUnit(vector<char>& unit) {
        vector<char> unitNums;
        unordered_set<char> unitSet;

        for (char square : unit) {
            if (isdigit(square)) unitNums.push_back(square);
        }
        
        for (char square : unitNums) {
            unitSet.insert(square);
        }
        
        return unitNums.size() == unitSet.size();
    }
};

File: Python/Arrays & Hashing/valid-sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        return self.is_row_valid(board) and \
        self.is_column_valid(board) and \
        self.is_subgrid_valid(board)

    def is_row_valid(self, board):
        for row in board:
            if not self.is_unit_valid(row):
                return False
        return True
    
    def is_column_valid(self, board):
        for column in zip(*board):
            if not self.is_unit_valid(column):
                return False
        return True
    
    def is_subgrid_valid(self, board):
        for i in (0, 3, 6):
            for j in (0, 3, 6):
                subgrid = [board[x][y] for x in range(i, i + 3) for y in range(j, j + 3)]
                if not self.is_unit_valid(subgrid):
                    return False
        return True
    
    def is_unit_valid(self, unit):
        unit = [square for square in unit if square != '.']
        return len(set(unit)) == len(unit)
